ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+#include <ruby.h>
+#include <stdio.h>
+
+#include <errno.h>
+#include <sys/types.h>
+#include <sys/ipc.h>
+#include <sys/sem.h>
+#include <sys/shm.h>
+#include <sys/time.h>
+
+#include <openssl/sha.h>
+#include <unistd.h>
+#include <stdbool.h>
+
+// needed for semctl
+union semun {
+  int              val;    /* Value for SETVAL */
+  struct semid_ds *buf;    /* Buffer for IPC_STAT, IPC_SET */
+  unsigned short  *array;  /* Array for GETALL, SETALL */
+  struct seminfo  *__buf;  /* Buffer for IPC_INFO
+                             (Linux-specific) */
+};
+
+#if defined(HAVE_RB_THREAD_CALL_WITHOUT_GVL) && defined(HAVE_RUBY_THREAD_H)
+// 2.0
+#include <ruby/thread.h>
+#define WITHOUT_GVL(fn,a,ubf,b) rb_thread_call_without_gvl((fn),(a),(ubf),(b))
+#elif defined(HAVE_RB_THREAD_BLOCKING_REGION)
+ // 1.9
+typedef VALUE (*my_blocking_fn_t)(void*);
+#define WITHOUT_GVL(fn,a,ubf,b) rb_thread_blocking_region((my_blocking_fn_t)(fn),(a),(ubf),(b))
+#endif
+
+// struct sembuf { // found in sys/sem.h
+//   unsigned short sem_num; /* semaphore number */
+//   short sem_op; /* semaphore operation */
+//   short sem_flg; /* operation flags */
+// };
+
+typedef struct {
+  int successes;
+  int arr_length;
+  double errors[];
```

Comment:
I'd rename those to `int counter; int window_length; double window[];`, in general I think it'd make sense to JUST have the C be a sliding window implementation and not couple it with the circuit breaker logic at all (just have it use it).

Thoughts @csfrancis?